pkg/commands/copy.go

```diff
@@ -197,21 +197,43 @@ func (cr *CachingCopyCommand) From() string {
 }
 
 func resolveIfSymlink(destPath string) (string, error) {
-	baseDir := filepath.Dir(destPath)
-	if info, err := os.Lstat(baseDir); err == nil {
+	if !filepath.IsAbs(destPath) {
+		return "", errors.New("dest path must be abs")
+	}
+
+	pathPrefix := "/"
+	pathTokens := strings.Split(destPath, string(os.PathSeparator))
+	for i, pathToken := range pathTokens {
```

Comment:
Can you explain the idea behind this code. It's not clear what it accomplishes.

Reply:
To resolve all symlinks in destination path.

The previous code failed to resolve unless symlink was a parent path. For example, when the destination is `/root/symlink/inner/target`, we expect it to be resolved to`/root/real/inner/target`, but previous code didn't work that way, it causing problems.
Similarly, if more than one symlink was found in the destination, it did not resolved correctly.
Problems also occur when symlink points to abs path.

So I split path into path elements and then re-joined them one by one. if symlinks is found in this process, it is replaced with a pointed path.
Then, if a path that does not exist, all remain path elements are need to be created, so they are just joined and returned.
This way, no matter how many symlinks are in the destination, no problem.

Reply:
Ah ok, thanks for that explanation. I think I better understand the problem now.

Would https://godoc.org/path/filepath#EvalSymlinks accomplish that goal?

Reply:
Yeah. At first I tried to solve this problem with `filepath.EvalSymlinks`. However, `file path.EvalSymlinks` could not be used because it returns a zero value string and error when` destPath` does not exist. We need to create a new directory even if `destPath` does not exist.

Reply:
Ah ok, that makes sense.

Would it be possible to use `EvalSymlink` by resolving the path from the longest to shortest segment?

Something like
```
for path != "/" {
  info, err := os.Lstat(path)
  if err != nil {
    if os.IsNotExist(err) {
      path = filepath.Dir(path)
      continue
    }
    // handle fatal err
  }
  path = path.EvalSymlinks(path)
}
```

WDYT?

Reply:
LGTM.

like this?
```go
	var nonexistentPaths []string

	newPath := destPath
	for newPath != "/" {
		_, err := os.Lstat(newPath)
		if err != nil {
			if os.IsNotExist(err) {
				dir, file := filepath.Split(newPath)
				newPath = filepath.Clean(dir)
				nonexistentPaths = append(nonexistentPaths, file)
				continue
			} else {
				return "", errors.Wrap(err, "failed to lstat")
			}
		}

		newPath, err = filepath.EvalSymlinks(newPath)
		if err != nil {
			return "", errors.Wrap(err, "failed to eval symlinks")
		}
		break
	}

	for i := len(nonexistentPaths)-1; i >= 0; i-- {
		newPath = filepath.Join(newPath, nonexistentPaths[i])
	}

	return filepath.Clean(newPath), nil

```

Reply:
Yup :)

Though a minor suggestion
```
for i := len(nonexistentPaths)-1; i >= 0; i-- {
  newPath = filepath.Join(newPath, nonexistentPaths[i])
}
```
to
```
sort.Reverse(nonexisitentPaths)
newPath = filepath.Join(newPath, nonexistentPaths...)
```

WDYT?

Reply:
In that case, the code actually looks like this:
```go
sort.Sort(sort.Reverse(sort.StringSlice(nonexistentPaths)))
newPath = filepath.Join(append([]string{newPath}, nonexistentPaths...)...)
```
Umm... I think it's a bit harder to read..